lucille-core/src/main/java/com/kmwllc/lucille/indexer/OpenSearchIndexer.java

```diff
@@ -92,46 +96,57 @@ protected void sendToIndex(List<Document> documents) throws Exception {
-      Map<String, Object> indexerDoc = getIndexerDoc(doc);
+      Map<String, Object> indexerDoc = doc.asMap();
 
       // remove children documents field from indexer doc (processed from doc by addChildren method call below)
       indexerDoc.remove(Document.CHILDREN_FIELD);
 
       // if a doc id override value exists, make sure it is used instead of pre-existing doc id
       String docId = Optional.ofNullable(getDocIdOverride(doc)).orElse(doc.getId());
+      indexerDoc.put(Document.ID_FIELD, docId);
 
       // handle special operations required to add children documents
       addChildren(doc, indexerDoc);
 
-      // create new IndexRequest
-      IndexRequest indexRequest = new IndexRequest(index);
-      indexRequest.id(docId);
-      if (routingField != null) {
-        indexRequest.routing(doc.getString(routingField));
-      }
-      indexRequest.source(indexerDoc);
-
-      if (versionType != null) {
-        indexRequest.versionType(versionType);
-        if (versionType == VersionType.EXTERNAL || versionType == VersionType.EXTERNAL_GTE) {
-          // the partition doesn’t need to be included in the version. We assume the doc id is used as the
-          // kafka message key, which should guarantee that all “versions” of a given document have the
-          // same kafka message key and therefore get mapped to the same topic partition.
-          indexRequest.version(((KafkaDocument) doc).getOffset());
-        }
+      String routing = doc.getString(routingField);
+      Long versionNum = (versionType == VersionType.External || versionType == VersionType.ExternalGte)
+          ? ((KafkaDocument) doc).getOffset()
+          : null;
+
+      if (update) {
+        br.operations(op -> op
+            .update((up) -> {
+              up.index(index).id(docId);
+              if (routingField != null) {
+                up.routing(routing);
+              }
+              if (versionNum != null) {
+                up.versionType(versionType).version(versionNum);
+              }
+              return up.document(indexerDoc);
+            }));
+      } else {
+        br.operations(op -> op
+            .index((up) -> {
+              up.index(index).id(docId);
+              if (routingField != null) {
+                up.routing(routing);
+              }
+              if (versionNum != null) {
+                up.versionType(versionType).version(versionNum);
+              }
+              return up.document(indexerDoc);
+            }));
       }
-
-      // add indexRequest to bulkRequest
-      bulkRequest.add(indexRequest);
     }
-
-    BulkResponse response = client.bulk(bulkRequest, RequestOptions.DEFAULT);
-    if (response.hasFailures()) {
-      log.error(response.buildFailureMessage());
+    BulkResponse response = client.bulk(br.build());
+    if (response != null && response.errors()) {
+      for (BulkResponseItem item : response.items()) {
+        throw new IndexerException(item.error().reason());
```

Comment:
we only want to throw an exception for the first item that has an error
so the throw should be inside a conditional that looks for the presence of an error

Reply:
Done